actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -26,7 +26,68 @@ module TagHelper
       PRE_CONTENT_STRINGS[:textarea]  = "\n"
       PRE_CONTENT_STRINGS["textarea"] = "\n"
 
+      class TagBuilder #:nodoc:
+        include TagHelper
 
+        VOID_ELEMENTS = %w(base  br  col  embed  hr  img  input  keygen  link
+                        meta  param  source  track  wbr).to_set
+
+        VOID_ELEMENTS.merge(VOID_ELEMENTS.map(&:to_sym))
+
+        ELEMENTS = %w(a abbr acronym address applet area article aside audio b
+                   base basefont bdi bdo bgsound big blink blockquote body br
+                   button canvas caption center cite code col colgroup command
+                   content data datalist dd del details dfn dialog dir div dl
+                   dt element em embed fieldset figcaption figure font footer
+                   form frame frameset h1 h2 h3 h4 h5 h6 head header hgroup hr
+                   html i iframe image img input ins isindex kbd keygen label
+                   legend li link listing main map mark marquee menu menuitem
+                   meta meter multicol nav nextid nobr noembed noframes
+                   noscript object ol optgroup option output p param picture
+                   plaintext pre progress q rp rt rtc ruby s samp script
+                   section select shadow small source spacer span strike strong
+                   style sub summary sup table tbody td template textarea tfoot
+                   th thead time title tr track tt u ul var video wbr xmp
+                   ).to_set
+
+        ELEMENTS.merge(ELEMENTS.map(&:to_sym))
+
+        def initialize(view_context)
+          @view_context = view_context
+        end
+
+        private
+
+          def tag_string(name, content_or_options = nil, options_or_escape = nil, escape = true, &block)
+            raise_if_void_tag_with_content(name, content_or_options, &block)
+            if block_given?
+              content_tag_string(name, @view_context.capture(self, &block), content_or_options, options_or_escape)
+            elsif content_or_options.is_a? String
+              content_tag_string(name, content_or_options, options_or_escape, escape)
+            elsif VOID_ELEMENTS.include?(name)
+              options_or_escape = true if options_or_escape.nil?
+              "<#{name}#{tag_options(content_or_options, options_or_escape)}>".html_safe
+            else
+              options_or_escape = true if options_or_escape.nil?
+              content_tag_string(name, "", content_or_options, options_or_escape)
+            end
+          end
+
+          def raise_if_void_tag_with_content(name, content_or_options, &block)
+            has_content = block_given? || content_or_options.is_a?(String)
+            void_with_content = has_content && VOID_ELEMENTS.include?(name)
+            raise ArgumentError, "Void tag with content" if void_with_content
+          end
```

Comment:
IMO, we don't need this protection. If someone wants to bastardize a BR tag with content, that's their prerogative. We just do the right thing by default.